---
title: "I. RNAseq analysis of Effect of Crotonic acid on OregonR flies 
description: "DGE analysis based on DESeq2"
principal investigator: "Eniola Adurosakin"
researchers: "Eniola Adurosakin, Joaquín de Navascués, Patrick Varga-Weisz"
output: 
  html_document:
    toc: true
    toc_float: true
    code_folding: show
    theme: readable
    df_print: paged
    css: doc.css
---
```{r setup, echo=FALSE, cache=FALSE}
ggplot2::theme_set(ggpubr::theme_pubr(base_size=10))
fsep <- .Platform$file.sep
knitr::opts_chunk$set(dev = 'png', 
                      fig.align = 'center', fig.height = 7, fig.width = 8.5, 
                      pdf.options(encoding = "ISOLatin9.enc"),
                      fig.path=paste0('notebook_figs', fsep), warning=FALSE, message=FALSE)
```


# 1 Preparation


**Libraries:**
```{r libraries, warning=FALSE, message=FALSE}
if (!require("librarian")) install.packages("librarian")
librarian::shelf(dplyr, stringr,
                 DESeq2, edgeR, biomaRt, readxl, rtracklayer, GenomicFeatures, limma,ggplot2, pheatmap, scico, dendsort,
                 here, writexl, gzcon,
                 quiet = TRUE)
```

**Set working directory:**
```{r setwd}
if (Sys.getenv("RSTUDIO")==1) {
   # setwd to where the editor is, if the IDE is RStudio
  setwd( dirname(rstudioapi::getSourceEditorContext(id = NULL)$path) )
} else {
  # setwd to where the editor is in a general way - maybe less failsafe than the previous
  setwd(here::here())
  # the following checks that the latter went well, but assuming
  # that the user has not changed the name of the repo
  d <- str_split(getwd(), fsep)[[1]][length(str_split(getwd(), fsep)[[1]])]
  if (d != 'Puigetal2023_bioinformatics_scripts') { stop(
    paste0("Could not set working directory automatically to where this",
           " script resides.\nPlease do `setwd()` manually"))
    }
}
```

**To save images outside the repo (to reduce size):**
```{r define_dir2figs}
## to keep it lightweight for GitHub:
# figdir <- paste0(c(head(str_split(getwd(), fsep)[[1]],-1),
#                    paste0(tail(str_split(getwd(), fsep)[[1]],1), '_figures')),
#                  collapse = fsep)
# dir.create(figdir, showWarnings = FALSE)

## for Zenodo:
figdir <- file.path(getwd(), 'figures')
dir.create(figdir, showWarnings = FALSE)
```


## 1.1 Experimental conditions

OregonR flies were treated with two concentrations of crotonic acid (0mM and 50mM). Data was collected in triplicates across five different time points 

## 1.2 Load raw count data
# 1.2.1 Read the full featureCounts table from Excel
```{r}
countsData <- read_excel("/Users/oluwapelumiadurosakin/Documents/Annotated Probe Report for All Probes.xls", sheet = 1)
head(countsData)

#Convert to data.frame for easier handling
countsData <- as.data.frame(countsData)

# Extract count matrix
rawData <- countsData[, 7:ncol(countsData)]   # all count columns
rownames(rawData) <- countsData$Gene_Id
head(rawData)

```

```{r}
# remove genes with low counts
cpms <- cpm(rawData)
keep <- rowSums(cpms > 1) >= 3 # detected in at least 3 samples
rawData <- rawData[keep,]
rawData <- rawData[order(row.names(rawData)), ]

# save
dirpath <- file.path(getwd(), 'output')
if ( !dir.exists(dirpath) ) dir.create( dirpath )
write.table(rawData, file = file.path(dirpath, 'featureCounts_allsamples.csv'),
            sep = '\t', row.names = TRUE, col.names = TRUE)
```


# 2 Differential gene expression
## 2.1 Set up `DESeq2DataSet` object
#Create experimental design object that contains the information from the metadata file:
# Build metadata
```{r}
#Read metadata from Excel
metadata <- read_excel("/Users/oluwapelumiadurosakin/Documents/metadata.xlsx")

# Convert to data.frame and set rownames to sample IDs
metadata <- as.data.frame(metadata)
rownames(metadata) <- metadata$Sample_ID
metadata$Sample_ID <- NULL  # optional: remove sampleID column

#Check matching with rawData columns
if (!all(colnames(rawData) %in% rownames(metadata))) {
  stop("Some samples in rawData are missing in metadata!")
}

# Reorder metadata to match rawData columns
metadata <- metadata[colnames(rawData), ]

#Convert relevant columns to factors
metadata$Time <- factor(metadata$Time, levels = c("T1","T2","T3","T4","T5"))
metadata$Condition <- factor(metadata$Condition, levels = c("Control","Treatment"))
```

```{r dge, warning=FALSE}
# Choose the design formula
# Option 1: Only treatment effect
exptObject <- DESeqDataSetFromMatrix(countData = rawData,
                              colData = metadata,
                              design =  ~ Condition)
# specify 'Control' as the reference level
exptObject$Condition <- relevel(exptObject$Condition, ref = "Control")
```

```{r}
# Option 2: Account for time points

exptObject2<- DESeqDataSetFromMatrix(countData = rawData,
                              colData = metadata,
                              design = ~ Time + Condition)
# specify 'T1' as the reference level
exptObject$Time <- relevel(exptObject$Time, ref = "T1")
```


## 2.2 Exploratory Data Analysis
Transform the normalized counts and plot them as PCA based on the Condition
```{r pca-raw}
vsd_Object <- vst(exptObject, blind=TRUE)
plotPCA(vsd_Object, intgroup = "Condition")
```


#To observe the effect of treatment across the time points 
```{r}
pcaData <- plotPCA(vsd_Object, intgroup = c("Condition", "Time"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))

ggplot(pcaData, aes(PC1, PC2, color = Condition, shape = Time)) +
  geom_point(size = 3) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  theme_minimal() +
  ggtitle("PCA of VST-normalized counts")
#observation:Treatment effect becomes weaker with respect to increasing time points. This is seen in the trajectory across time points, indicating that time is a major driver of the expression changes
```

#Independent plots for the different time points
```{r}
pcaData <- plotPCA(vsd_Object, intgroup = c("Condition", "Time"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color = Condition)) +
  geom_point(size = 3) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  facet_wrap(~ Time) +
  theme_minimal() +
  ggtitle("PCA by Time Point (Control vs Treatment)")

```

 So we save this:
```{r save-vst-bcorr-counts}
saveRDS(vsd_Object, file.path(dirpath, 'vst_pseudocounts_batchCorrected.RDS'))
```


### Sample correlations
To further visualise the similarity of the samples we use the transformed, normalized counts and perform hierarchical clustering (with hidden dendrograms):
```{r}
# correlation matrix
vsd_cor_Object <- cor(assay(vsd_Object))

# sample annotations from metadata
sample_annots <- metadata[, c("Condition", "Time")]

# heatmap
pheatmap(
  mat               = vsd_cor_Object,
  scale             = "none",
  cellwidth         = 15,
  cellheight        = 15,
  main              = "RNA-seq Sample Correlations",
  fontsize          = 12,
  annotation        = sample_annots,   # annotate by condition + time
  cluster_rows      = TRUE,
  cluster_cols      = TRUE,
  show_rownames     = TRUE,
  labels_row        = rownames(metadata),
  show_colnames     = TRUE,
  labels_col        = rownames(metadata),
  color             = scico(255, palette = "bamako"),
  border_color      = "grey80"
)

```

This is somewhat redundant with PCA but also makes sense, so we will save the `vsd_Object` for the descriptive visualisations.
```{r save-pseudocounts}
saveRDS(vsd_Object, file.path(dirpath, 'vsd.RDS'))
```


## 2.3 Visualise dispersion estimates


This will normalise the data, correct for dispersion (variance between replicates) and set data up for a differential comparison of any 2 conditions:
```{r analysis-object}
analysisObject = DESeq(exptObject)
```

After fitting the data to a negative binomial model, some basic QC consists of plotting the dispersion estimates using the `plotDispEsts()` function. The dispersion estimates are used to model the raw counts, assuming (1) that dispersions will generally decrease with increasing mean, and (2) that they should more or less follow the fitted line.
```{r plot-dispersion}
plotDispEsts(analysisObject)
```

This seems reasonably in order, so we move on to saving the expression data for further visualisation and analysis, and Supplementary Data:
```{r save-counts}
rawCounts        <- as.data.frame(counts(analysisObject, normalized=FALSE))
normalisedCounts <- as.data.frame(counts(analysisObject, normalized=TRUE))
saveRDS(rawCounts, file.path(dirpath, 'rawCounts.RDS'))
saveRDS(normalisedCounts, file.path(dirpath, 'normalisedCounts.RDS'))
```


## 2.4 Differential gene expression
Loop over the experimental conditions and save `DESeq2::results` as RDS data:
```{r}
# conditions to be tested (everything except Control)
test_conditions <- unique(exptObject$Condition[exptObject$Condition != "Control"])

# create a list to store results
tests <- list()

for (condtn in test_conditions) {
  # get DESeq2 results comparing condition vs Control
  deData <- as.data.frame(results(
    analysisObject,
    contrast = c("Condition", condtn, "Control"),  # test vs reference
    pAdjustMethod = "BH"
  ))
  
  # add gene IDs from rownames
  deData <- tibble::rownames_to_column(deData, var = "geneID")
  
  # sort by p-value
  deData <- deData[order(deData$pvalue), ]
  
  # save as RDS
  saveRDS(deData, file = file.path(dirpath, paste0("Control_vs_", condtn, ".RDS")))
  
  # keep simplified version (geneID, log2FC, padj) in results list
  cols <- c("geneID", "log2FoldChange", "padj")
  tests[[paste0("Control_vs_", condtn)]] <- deData[, cols]
}
```


# to get a more informative naming for the samples:
targets$sampleIDs <- names(rawCounts)
# conditions to be tested
test_conditions <- unique( targets[targets$Condition != 'Control',]$Condition )
test_names <- paste0(rep('Control_vs_',length(test_conditions)),test_conditions)
tests <- as.list(rep(NA, length(test_names)))
names(tests) <- test_names 

for (condtn in test_conditions) {
  # get the Counts for those conditions
  deData <- as.data.frame(results(analysisObject,
                                  contrast=c("condition", condtn, 'Control'), # Reference goes last!
                                  pAdjustMethod="BH"))
  # add column of ID
  deData <- cbind(data.frame('ensemblGeneID'=rownames(deData)), deData)
  # sort by pval
  deData <- deData[order(deData$pvalue), ] 
  # add gene symbol column and reorder columns
  deData <- merge(deData, dlist, by=0)
  deData <- deData[,c(1,ncol(deData),2:(ncol(deData)-1))]
  # save for later
  saveRDS(deData, file=file.path(dirpath, paste0('Control_vs_', condtn, ".RDS")))
  # save as Supplementary data for publication
  cols <- c('gene_symbol', 'ensemblGeneID', 'log2FoldChange', 'padj')
  rownames(deData) <- NULL
  tests[[paste0('Control_vs_', condtn)]] <- deData %>% dplyr::select(all_of(cols))
}
```
(Note that in the parameter `contrast`, `'Control'` is last - I cannot find formal justification of this, but even with the `relevel`ling of the `condition` column of the experimental design object to make 'Control' the reference level, this needs to go last, or the log~2~FC values will have the opposite sign.)


### Supplementary Table S3


Now we just need to write the adjusted p-values and the log~2~(fold changes). That, with the raw counts, should be enough for anyone to do their own 'light' analysis without having to get any files from GEO.
```{r table-s3}
# add `rawCounts` to `tests`
testsappendage <- rawCounts %>% tibble::rownames_to_column(var = "ensembl_id")
tests <- rlist::list.append(tests, `Raw counts per gene per sample` = testsappendage)
# Supplementary data for publication
write_xlsx(tests, path = file.path('output', 'Table S3.xlsx'))
```

Finally, to have the experimental designed captured in a flexible manner for later on:
```{r save-targets}
targets$condition_md <- plyr::mapvalues(
  targets$Condition,
  from=unique(targets$Condition),
  to=c('*da^RNAi^*', '*da*', '*wild-type*', '*da:da*', '*scute*')
  )
targets$condition_md <- factor(
  targets$condition_md,
  c('*wild-type*', '*da*', '*da:da*','*da^RNAi^*', '*scute*')
  )
saveRDS(targets, file.path(dirpath, 'targets.RDS'))
```
